specs/MultiProfile.md

```diff
@@ -165,6 +165,28 @@ void ScenarioCookieManagement::DeleteAllCookies()
     CHECK_FAILURE(m_cookieManager->DeleteAllCookies();
 }
 ```
+
+### Delete profile
+
+```cpp
+HRESULT AppWindow::DeleteProfile(ICoreWebView2Controller* controller)
+{
+    wil::com_ptr<ICoreWebView2> coreWebView2;
+    CHECK_FAILURE(controller->get_CoreWebView2(&coreWebView2));
+    auto webview7 = coreWebView2.try_query<ICoreWebView2_7>();
+    if (webview7)
+    {
+        wil::com_ptr<ICoreWebView2Profile> profile;
+        CHECK_FAILURE(webview7->get_Profile(&profile));
+        auto profile2 = profile.try_query<ICoreWebView2StagingProfile4>;
+        if (profile2)
+        {
+            CHECK_FAILURE(profile2->Delete());
```

Comment:
The WV2s who under the profile will receive `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED `message and will be closed while the profile is deleted.
About **If somebody deletes it out from under you, then profile methods try to act inert.**

- If Methods are called before the profile delete, work will as normal
- If the WV2 has received `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED `message, the user should not call these methods.
- If Methods are called between profile delete and received `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED `message, the get_Methods will return last-known-good values, because it does not ask browser process data just return client data. And the put_Methods will be failure because the MOJOM has closed.

About **What if a new profile is created in the same directory. Do we now have two profiles with the same ProfilePath?**
Every profile has its own directory named with prefix WV2Profile_ProfileName, if the profile is deleted, it will be marked as deleted and will be deleted from disk when browser process exited. During this time, user could not create profile with the same name. 

About **Willing to let these methods throw errors, since the app is the one that deleted them. But inert behavior is also acceptable.**
If the one has deleted the profile and then I try to delete it, if I has received `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED `message, I should not do this.
if not receive, the funtion will successful. It will always successful. Just like that we set a flag to true, and then we can set it to true again although the flag has been set to true.

Reply:
> If the WV2 has received `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message, the user should not call these methods.

1. What happens if the app calls the methods after receiving the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message?
2. Old apps do not understand the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message. They will just ignore it, and continue to use the profile. Does that mean that old apps will start crashing if the profile is deleted out from under them?

Suppose Excel and Word both use the same profile. Both are using the Evergreen version, so they have been silently upgraded to the latest WebView2. Word gets an update that will delete profiles after 1 week (to avoid accumulating cruft). Excel hasn't been updated yet. Word deletes a profile. Excel gets the `DELETED` message but doesn't know what it means (since it hasn't been updated yet). Excel tries to use the profile anyway. What happens? Does Excel crash? Who is at fault here?

* Is it Excel's fault for continuing to use a deleted profile (even though the concept of a "deleted profile" didn't exist at the time Excel was written)?
* Is it Word's fault for deleting a profile before everybody who was sharing the profile has been upgraded to understand profile deletion?

Reply:
Could we rename the profile folder to wv2profile_deleted# and then mark it for deletion so that that profile name is available for use immediately?

If the profile is deleted I thought I understood from you that the CoreWebView2 is closed in the same manner as if CoreWebView2Controller.Close were called, or if the browser process crashed. We should make use of the existing close mechanism and we should document this by linking to existing Close documentation that describes this.

If two apps have decided to take the big step of joining together and sharing a user data folder they must be prepared for the consequences. In the case of the Office example perhaps they could delay deleting profiles until all apps are updated. Or version their profiles. Let's add documentation something like "Profiles are shared amongst all apps sharing the same user data folder. Accordingly, deleting a profile affects all apps sharing a user data folder. It is up to the apps to ensure they do not break one another by deleting profiles in use by another app."

Reply:
> > If the WV2 has received `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message, the user should not call these methods.
> 
> 1. What happens if the app calls the methods after receiving the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message?
> 2. Old apps do not understand the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message. They will just ignore it, and continue to use the profile. Does that mean that old apps will start crashing if the profile is deleted out from under them?
> 
> Suppose Excel and Word both use the same profile. Both are using the Evergreen version, so they have been silently upgraded to the latest WebView2. Word gets an update that will delete profiles after 1 week (to avoid accumulating cruft). Excel hasn't been updated yet. Word deletes a profile. Excel gets the `DELETED` message but doesn't know what it means (since it hasn't been updated yet). Excel tries to use the profile anyway. What happens? Does Excel crash? Who is at fault here?
> 
> * Is it Excel's fault for continuing to use a deleted profile (even though the concept of a "deleted profile" didn't exist at the time Excel was written)?
> * Is it Word's fault for deleting a profile before everybody who was sharing the profile has been upgraded to understand profile deletion?
**1. What happens if the app calls the methods after receiving the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message?**
We have a guard in client to, when the message receiving , the guard flag will be set, so every API call will fail.
**Old apps do not understand the `COREWEBVIEW2_PROCESS_FAILED_REASON_PROFILE_DELETED` message** Dot worry about, multiple profile means the WV2s work in same userfolder, so them share a same browser. So, the old app will run under the same version with other app. In other word, if other app upgrade, the old app must to upgrade otherwise its clint's version will not adapt the browser.

Reply:
> Could we rename the profile folder to wv2profile_deleted# and then mark it for deletion so that that profile name is available for use immediately?
> 
> If the profile is deleted I thought I understood from you that the CoreWebView2 is closed in the same manner as if CoreWebView2Controller.Close were called, or if the browser process crashed. We should make use of the existing close mechanism and we should document this by linking to existing Close documentation that describes this.
> 
> If two apps have decided to take the big step of joining together and sharing a user data folder they must be prepared for the consequences. In the case of the Office example perhaps they could delay deleting profiles until all apps are updated. Or version their profiles. Let's add documentation something like "Profiles are shared amongst all apps sharing the same user data folder. Accordingly, deleting a profile affects all apps sharing a user data folder. It is up to the apps to ensure they do not break one another by deleting profiles in use by another app."

I've thought about this. If a file is occupied, it still can rename and mark it for deletion. But unfortunate, for a folder we cannot do like this. The rename will be failure if any file in the folder is occupied.